packaging/_parser.py

```diff
@@ -0,0 +1,155 @@
+            marker += tokens.read().text
+    else:
+        marker = ""
+        tokens.expect("stringEnd")
+    return (name, url, extras, specifier, marker)
+
+
+def parse_extras(tokens: Tokenizer) -> List[str]:
+    """
+    EXTRAS: (LBRACKET + IDENTIFIER + (COLON + IDENTIFIER)* + RBRACKET)*
+    """
+    extras = []
+    if tokens.try_read("LBRACKET"):
+        while tokens.match("IDENTIFIER"):
+            extras.append(tokens.read("IDENTIFIER").text)
+            tokens.try_read("COLON")
+        if not tokens.try_read("RBRACKET"):
+            tokens.raise_syntax_error("Closing square bracket is missing")
+    return extras
+
+
+def parse_specifier(tokens: Tokenizer) -> str:
+    """
+    SPECIFIER: LPAREN (OP + VERSION + COLON)+ RPAREN | OP + VERSION
+    """
+    parsed_specifiers = ""
+    lparen = False
+    if tokens.try_read("LPAREN"):
+        lparen = True
+    while tokens.match("OP"):
+        parsed_specifiers += tokens.read("OP").text
+        if tokens.match("VERSION"):
+            parsed_specifiers += tokens.read("VERSION").text
+        else:
+            tokens.raise_syntax_error("Missing version")
+        if tokens.match("COLON"):
+            parsed_specifiers += tokens.read("COLON").text
+    if lparen and not tokens.try_read("RPAREN"):
+        tokens.raise_syntax_error("Closing right parenthesis is missing")
+    return parsed_specifiers
+
+
+def parse_quoted_marker(tokens: Tokenizer) -> List[str]:
+    tokens.try_read("SEMICOLON")
+    return parse_marker_expr(tokens)
+
+
+def parse_marker_expr(tokens: Tokenizer) -> List[str]:
+    """
+    MARKER_EXPR: MARKER_ATOM (BOOLOP + MARKER_ATOM)+
+    """
+    expression = [parse_marker_atom(tokens)]
+    while tokens.match("BOOLOP"):
+        tok = tokens.try_read("BOOLOP")
+        expr_right = parse_marker_atom(tokens)
+        expression.extend((tok.text, expr_right))
+    return expression
+
+
+def parse_marker_atom(tokens: Tokenizer) -> Any:
+    """
+    MARKER_ATOM: LPAREN MARKER_EXPR RPAREN | MARKER_ITEM
+    """
+    if tokens.try_read("LPAREN"):
+        marker = parse_marker_expr(tokens)
+        if not tokens.try_read("RPAREN"):
+            tokens.raise_syntax_error("Closing right parenthesis is missing")
+        return marker
+    else:
+        return parse_marker_item(tokens)
+
+
+def parse_marker_item(tokens: Tokenizer) -> Tuple[Any, Any, Any]:
+    """
+    MARKER_ITEM: MARKER_VAR MARKER_OP MARKER_VAR
+    """
+    marker_var_left = parse_marker_var(tokens)
+    marker_op = parse_marker_op(tokens)
+    marker_var_right = parse_marker_var(tokens)
+    return (marker_var_left, marker_op, marker_var_right)
+
+
+def parse_marker_var(tokens: Tokenizer) -> Any:
+    """
+    MARKER_VAR: VARIABLE MARKER_VALUE
+    """
+    if tokens.match("VARIABLE"):
+        return parse_variable(tokens)
+    else:
+        return parse_python_str(tokens)
+
+
+def parse_variable(tokens: Tokenizer) -> Any:
+    from .markers import Variable
```

Comment:
We should move the AST classes to over here, allowing them to be used here without needing a deferred import to avoid import cycles.